### ISSUE TITLE
Document, how does NNCF compute gradients for QAT

### ISSUE THREAD

**Issue by ruro** [openvinotoolkit/nncf#3870]
I was trying to figure out, what does NNCF actually do in order to make the various fake quantization functions differentiable.

The original "[Neural Network Compression Framework for fast model inference](https://arxiv.org/pdf/2002.08679)" paper claimed that

>  ... methods proposed in [12], where quantization parameters are learned using gradient descent. In our framework we use a similar quantization method, along with other quantization schemes, while also providing the ability to automatically insert Fake Quantization operations in the model graph.

where reference 12 is "[PACT: Parameterized Clipping Activation for Quantized Neural Networks](https://arxiv.org/pdf/1805.06085)".

However, it seemingly doesn't disclose which "similar quantization method" it actually uses. It's not really PACT, since PACT doesn't support "proper" asymmetric quantization, requires replacing ReLU with a custom activation function and adds an extra regularization term into the loss.

Unlike PACT, NNCF just defines custom `forward`/`backward` functions that are differentiable wrt `inputs`, `inputs_low` **and** `inputs_range` (without relying on custom activation functions or extra loss terms). The implementation for the custom `forward`/`backward` functions in question can be found in

- `src/nncf/torch/quantization/reference.py` (pure python implementation)
- `src/nncf/torch/extensions/src/quantization/cpu/functions_cpu.cpp` (CPU)
- `src/nncf/torch/extensions/src/quantization/cuda/functions_cuda_impl.cu` (CUDA)

I think that the most similar paper to what NNCF does is actually "[Learned Step Size Quantization](https://arxiv.org/pdf/1902.08153)" (aka LSQ), not PACT. Although NNCFs implementation doesn't **quite** match with LSQ.

**Comment by Wan-Intel** [comment]
A uniform "fake" quantization method supports an arbitrary number of bits (>=2) which is used to represent weights and activations. The method performs differentiable sampling of the continuous signal (for example, activations or weights) during forward pass, simulating inference with integer arithmetic.]

For more information, please refer to [Uniform Quantization with Fine-Tuning](https://github.com/openvinotoolkit/nncf/blob/develop/docs/usage/Quantization.md#uniform-quantization-with-fine-tuning) in [OpenVINO™ Neural Network Compression Framework (NNCF)](https://github.com/openvinotoolkit/nncf/tree/develop).

**Comment by ruro** [comment]
@Wan-Intel I have read the documentation you linked, it doesn't answer my questions. The formulas described on that page aren't differentiable and the page doesn't describe, how do you define the (fake) gradients for these non-differentiable operations.

From reading the code, the gradients wrt `input` seem to be obtained by assuming that non-differentiable operations act like straight-through estimators (which is fairly common for fake quantization implementations at this point). However, the STE trick doesn't work for `input_low` / `input_range` and as I've mentioned, your implementation seems to do something vaguely similar to "[Learned Step Size Quantization](https://arxiv.org/pdf/1902.08153)", but it's not QUITE the same as LSQ.

**Comment by Wan-Intel** [comment]
Hi @ruro 
Let me check with relevant team, and we will provide an update soon.

**Comment by jgespino** [comment]
Ref. 181547

**Comment by ljaljushkin** [comment]
Thanks for the detailed analysis, @ruro!

NNCF employs a learned-range fake quantization method where input_low and input_range are learned via gradient descent. It is indeed closer to LSQ than to PACT.

The in-range gradient for input_range is mathematically equivalent to LSQ's step-size gradient, just expressed in a different parameterization.

We've added documentation for these gradient formulas to the Quantization.md page — see the new "Gradient Computation for Quantization-Aware Training" section (PR #3965).
